### PR TITLE
fix: vis is rendered with an invalid config

### PR DESCRIPTION
**Pull request by oltionchampari**
### Developer Checklist (Definition of Done)

**Issue**

- [ ] All acceptance criteria from the issue are met
- [x] Tested in latest Chrome/Firefox

**UI/UX/Vis**

- [ ] Requires UI/UX/Vis review
  - [ ] Reviewer(s) are notified (_tag assignees_)
  - [ ] Review has occurred (_link to notes_)
  - [ ] Feedback is included in this PR
  - [ ] Reviewer(s) approve of concept and design

**Code**

- [x] Branch is up-to-date with the branch to be merged with, i.e., develop
- [x] Code is cleaned up and formatted
- [ ] Unit tests are written (frontend/backend if applicable)
- [ ] Integration tests are written (if applicable)

**PR**

- [x] Descriptive title for this pull request is provided (will be used for release notes later)
- [x] Reviewer and assignees are defined
- [x] Add type label (e.g., *bug*, *feature*) to this pull request
- [x] Add release label (e.g., `release: minor`) to this PR following [semver](https://semver.org/)
- [x] The PR is connected to the corresponding issue (via `Closes #...`)
- [x] [Summary of changes](#summary-of-changes) is written

### Summary of changes
 We have the issue that we are providing vis an externalConfig that does not include all possible defaults
```typescript
  const config = {
    type: 'Heatmap plot',
    catColumnsSelected: [
      { id: 'Column 1', name: 'Column 1' },
      { id: 'Column 2', name: 'Column 2' },
    ],
  };
```
This is the different values the vis was being rendered with (causes errors) when inputing the above config
```
MainApp.tsx: Parent {type: 'Heatmap plot', catColumnsSelected: Array(2)}
MainApp.tsx:40 Parent null
MainApp.tsx:40 Parent {type: 'Heatmap plot', catColumnsSelected: Array(2), color: {…}}
MainApp.tsx:40 Parent {type: 'Heatmap plot', color: {…}, catColumnsSelected: Array(2), numColorScaleType: 'Sequential', xSortedBy: 'CAT_ASC', …}
```

And this with this fix

```
MainApp.tsx: Parent {type: 'Heatmap plot', catColumnsSelected: Array(2)}
MainApp.tsx:40 Parent {type: 'Heatmap plot', color: {…}, catColumnsSelected: Array(2), numColorScaleType: 'Sequential', xSortedBy: 'CAT_ASC', …}
```

We could improve the logic and readabilty of this section of code when we switch completely to mantine 7 to use https://mantine.dev/hooks/use-uncontrolled/

### Screenshots

before
gr-original
after
gr-new

### Additional notes for the reviewer(s)

-  
Thanks for creating this pull request 🤗

### CODE REVIEW

File: src/vis/heatmap/HeatmapGrid.tsx [hunk 1]
```diff
@@ -34,11 +34,8 @@ export function HeatmapGrid({
 
   return (
     <Stack align="center" justify="center" sx={{ width: '100%', height: '100%' }} p="sm">
-      {status === 'pending' ? (
-        <Loader />
-      ) : !hasAtLeast2CatCols ? (
-        <InvalidCols headerMessage="Invalid settings" bodyMessage="To create a heatmap chart, select at least 2 categorical columns." />
-      ) : (
+      {status === 'pending' && <Loader />}
```
Comment (oltionchampari) [inline src/vis/heatmap/HeatmapGrid.tsx:37]: The Invalid settings message is shown shortly while the state is `idle`. This change fixes that.

File: src/vis/heatmap/HeatmapText.tsx [hunk 2]
```diff
@@ -26,7 +26,7 @@ export function HeatmapText({
   isImmediate: boolean;
 }) {
   const labelSpacing = useMemo(() => {
-    const maxLabelLength = d3.max(yScale.domain().map((m) => m.length));
+    const maxLabelLength = d3.max(yScale.domain().map((m) => m?.length));
```
Comment (oltionchampari) [inline src/vis/heatmap/HeatmapText.tsx:29]: Accounts for null categories

**Comment by puehringer** [comment]
@oltionchampari regarding your last point of using useUncontrolled when we are using mantine7. develop is currently using mantine7, and mantine6 has the hook as well. So why don't we do it right away? 🤔

**Comment by oltionchampari** [comment]
> @oltionchampari regarding your last point of using useUncontrolled when we are using mantine7. develop is currently using mantine7, and mantine6 has the hook as well. So why don't we do it right away? 🤔

I thought it was in 7 and we would need the change soonish in v7.0.3. But since its in six already we can go ahead and change it already

**Comment by puehringer** [comment]
> > > > @oltionchampari regarding your last point of using useUncontrolled when we are using mantine7. develop is currently using mantine7, and mantine6 has the hook as well. So why don't we do it right away? 🤔
> > > 
> > > 
> > > I thought it was in 7 and we would need the change soonish in v7.0.3. But since its in six already we can go ahead and change it already
> > 
> > 
> > Ok, now I understand: this PR is intended to be released for a previous visyn_core version, right? But then the `develop` PR target isn't really correct, as that is latest (Mantine 7 only).
> 
> We will cherry pick this commit and merge it to v7.0.3 as far as i understand.

Yeah, that makes sense. I would still like to see the `useUncontrolled` in case you have time to add it. But thanks for the PR!

**Comment by oltionchampari** [comment]
@puehringer Please take another look when possible. 
Note the new screencasts of interactions before and current.
1. One notable difference now is that once the `EagerVis` receives and external config we merge the defaults to it but we do not update the config of the parent, that will occur once the user changes a config in the sidebar. This saves a couple of updates.
2. Note the short flickering of the type input is gone in this branch

File: src/vis/heatmap/HeatmapGrid.tsx (hunk 1 above)
Comment (thinkh) [reply src/vis/heatmap/HeatmapGrid.tsx:37]: Thanks for fixing this issue and the mention.

**Review by thinkh — APPROVED** [review]
Thanks for addressing the suggestion and updating the example. It looks good to me.

@puehringer please take a final look. Thanks.

**Pull request merged**